New search bar from WPMediaPicker in media library and editor picker

etoledom opened the pull request:
Fixes #7680 

This PR implements the new search bar from WPMediaPicker into MediaLibraryViewController and AztecPostViewController.

For the `MediaLibraryViewController`, the strategy used was to convert this controller into a `WPMediaPickerViewController` subclass. This way there is no need to layout the picker collection view manually.

For the `AztecPostViewController` the search bar was added through the Media Picker delegate methods. To match the media library behavior, the no-resutls view was also added and configured accordingly.

###Known issues:
- The `Cancel` button in the search bar is not styled correctly and remains of white color.

<img width="351" alt="screen shot 2017-11-20 at 12 26 28 am" src="https://user-images.githubusercontent.com/9772967/33000960-cd72159c-cd89-11e7-8de0-6d776deb6fb0.png">

To test:

- Run pod install to update the Media Picker library.
- Run the WPiOS App.
- Go to the Media library and test the search bar behavior.
- Go to post / page editors.
- Select add media from WordPress library (`w` button)
- Test the search bar behavior.

@SergioEstevao needs review.

etoledom commented:
I found this small issue when searching, dismissing and opening again the picker from the editor, it show the result from the previous search.

<img width="955" alt="screen shot 2017-11-21 at 10 34 08 am" src="https://user-images.githubusercontent.com/9772967/33075288-b5a3eba6-cea7-11e7-8d46-b3a1af3326e5.png">

SergioEstevao commented:
@etoledom it's looking good are you adding the search bar  to the icon Site Picker and featured image picker on this PR too?

SergioEstevao commented on WordPress/Classes/Extensions/WPNoResultsView+MediaPicker.swift:
```diff
@@ -0,0 +1,44 @@
+//
```
We normally don't use file header comments in your files, because they quickly became outdated.

SergioEstevao commented on WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift:
```diff
@@ -445,6 +445,16 @@ class AztecPostViewController: UIViewController, PostEditor {
+    ///
+    fileprivate let noResultsView: WPNoResultsView = {
+        let noResultsView = WPNoResultsView()
+        noResultsView.accessoryView = UIImageView(image: UIImage(named: "media-no-results"))
```
This configuration can probably be added to the Media extension  for WPNoResultsView has a convenience init. no?

etoledom replied:
Good point!

etoledom commented:
@SergioEstevao  thanks!

Right! I was focusing more on the Media Library and the Editors, but I see no problem with adding those search bars in the next commit as well. 👍

etoledom commented:
- Fixed the issue I mentioned about the search query not clearing on dismiss.
- Added a search bar and a no results view to the `SiteIconPicker`.
- Added a no result view to the `GravatarPicker`. I didn't add the search bar there since there is no WordPress media showing, only the device images.

I merged dev into this branch yesterday to avoid merge conflicts. Nonetheless there were two conflicts with commits from today 😱, but since they were simple, I resolved them.

SergioEstevao commented:
@frosty do you mind to give a look to this PR?

SergioEstevao commented:
@etoledom this is looking good the only place I don't see search activated is the setting of a Featured Image in Post settings.

frosty commented on WordPress/Classes/ViewRelated/Media/MediaLibraryViewController.swift:
```diff
@@ -172,116 +131,10 @@ class MediaLibraryViewController: UIViewController {
-
-        UIView.animate(withDuration: duration) {
-            self.noResultsView?.centerInSuperview()
+        if self.searchBar?.isFirstResponder == .some(true) {
```
You don't need `self.` here, and you can also leave out the `.some` – just `if ... == true`

frosty commented on WordPress/Classes/ViewRelated/Media/MediaLibraryViewController.swift:
```diff
@@ -508,16 +330,16 @@ class MediaLibraryViewController: UIViewController {
+        self.options = options
 
-        pickerViewController.clearSelectedAssets(true)
+        self.clearSelectedAssets(true)
```
Don't need `self.` here.

frosty commented on WordPress/Classes/ViewRelated/Media/MediaLibraryViewController.swift:
```diff
@@ -508,16 +330,16 @@ class MediaLibraryViewController: UIViewController {
     @objc private func trashTapped() {
         let message: String
-        if pickerViewController.selectedAssets.count == 1 {
+        if self.selectedAssets.count == 1 {
```
Don't need `self.` here

frosty commented on WordPress/Classes/ViewRelated/Media/MediaLibraryViewController.swift:
```diff
@@ -535,8 +357,8 @@ class MediaLibraryViewController: UIViewController {
     private func deleteSelectedItems() {
-        guard pickerViewController.selectedAssets.count > 0 else { return }
-        guard let assets = pickerViewController.selectedAssets as? [Media] else { return }
+        guard self.selectedAssets.count > 0 else { return }
```
Don't need `self.` here or in the line below.

frosty commented on WordPress/Classes/ViewRelated/Media/MediaLibraryViewController.swift:
```diff
@@ -882,7 +682,7 @@ extension MediaLibraryViewController: WPMediaPickerViewControllerDelegate {
             // of changes to the data source before the collection view has
             // updated its selected assets.
-            guard let assets = (pickerViewController.selectedAssets as? [Media]) else { return }
+            guard let assets = (self.selectedAssets as? [Media]) else { return }
```
Don't need `self.` here

frosty commented on WordPress/Classes/ViewRelated/Media/MediaLibraryViewController.swift:
```diff
@@ -931,31 +731,27 @@ extension MediaLibraryViewController: WPMediaPickerViewControllerDelegate {
         if !mediaProgressCoordinator.isRunning && pickerDataSource.isPaused {
             pickerDataSource.isPaused = false
-            pickerViewController.collectionView?.reloadData()
+            self.collectionView?.reloadData()
```
Don't need `self.` here

etoledom commented on WordPress/Classes/Extensions/WPNoResultsView+MediaPicker.swift:
```diff
@@ -0,0 +1,43 @@
+
+import Foundation
+
+extension WPNoResultsView {
```
Yes, the first idea was a subclass but I had troubles overriding the init method with the current declaration of the `WPNoResultsView` class (it forces to call initWithFrame:) so I made the extension with the config methods. Later on we decided to add the factory method, but I agree that the subclass with the factory method is still better :)

etoledom commented on WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift:
```diff
@@ -863,6 +869,34 @@ class AztecPostViewController: UIViewController, PostEditor {
     }
 
+    fileprivate func updateSearchBar(mediaPicker: WPMediaPickerViewController) {
+        if mediaPicker.dataSource?.numberOfAssets() == .some(0) {
```
I have been working on this part taking into account the search status too:
```
        let isSearching = mediaLibraryDataSource.searchQuery?.count ?? 0 != 0
        let hasAssets = mediaLibraryDataSource.numberOfAssets() > 0

        if isSearching || hasAssets {
```

`... .count ?? 0 != 0` has to be true when `searchQuery` is `nil`

etoledom commented:
Thanks @frosty for the review!

I found the reason why the cancel button color won't change:

This part of the customizeAppearance in the app delegate:
```   
    // Media Picker styles
    UIBarButtonItem *barButtonItem = [UIBarButtonItem appearanceWhenContainedInInstancesOfClasses:@[ [WPMediaPickerViewController class] ]];
    [barButtonItem setTitleTextAttributes:@{NSForegroundColorAttributeName: [UIColor whiteColor], NSFontAttributeName : [WPFontManager systemSemiBoldFontOfSize:16.0]} forState:UIControlStateNormal];
```

Is overriding the global style of the search bar inside the picker, since that cancel button is also an `UIBarButtonItem` inside the `WPMediaPickerViewController` class. For some reason this setting takes precedence over the specific `UISearchBar` appearance configuration.

@frosty or @SergioEstevao do you know if those settings are there for some reason in particular?
If it is, I can try to find another way to fix this, but removing it solves the cancel button color problem and all other bar buttons in the picker seem to be unchanged.

---
>  I guess we should probably just detect that the album is empty and not even show it in the first place...

If there are no assets to pick, the user still has the opportunity to take a selfie with the camera (on a device), so probably is good to show it anyway.

I agree that the empty message could be improved, about that I have a of questions: 
- What is the procedure to change displayed text, taking into account all the translations?

SergioEstevao commented:
> @frosty or @SergioEstevao do you know if those settings are there for some reason in particular?
> If it is, I can try to find another way to fix this, but removing it solves the cancel button color problem and all other bar buttons in the picker seem to be unchanged.

I think this appearance is set because of the buttons on the navigation toolbar when we select assets, and we have sentences like `Insert 1` `Insert 2`. This can be seen when selecting assets using the Aztec editor.

etoledom commented:
I see, it seems to be specific for the picker as inputView.

<img width="638" alt="screen shot 2017-11-24 at 2 29 44 pm" src="https://user-images.githubusercontent.com/9772967/33220213-345002b2-d125-11e7-96b1-1bfb2be15209.png">

etoledom commented:
Just pushed a new commit:
- Added the search bar to the featured image picker.
- Fixed the search bar styles.
- Applied the requested changes.

SergioEstevao commented:
@etoledom 

> What is the procedure to change displayed text, taking into account all the translations?

You just need to change it on the code, then our translations scripts will pick it up, and it will be send to our translators, before a new release is made we make sure that we update the translation strings on the app.

frosty commented on WordPress/Classes/ViewRelated/Media/MediaNoResultsView.swift:
```diff
@@ -1,15 +1,16 @@
-        let noResultView = WPNoResultsView()
-        noResultView.updateForNoMediaAssets(userCanUploadMedia: false)
+    @objc(view)
+    class func makeView() -> MediaNoResultsView {
```
Now this is a subclass, can this just be an `init` method?

etoledom commented:
@SergioEstevao @frosty Thank you both for your help!
I just sent a new commit addressing both of the issues mentioned before.

As always, I welcome any feedback you might have.

SergioEstevao commented:
@etoledom great work on those final fixes, @frosty do you want to give it a final look?

SergioEstevao commented:
@etoledom you may need to merge the latest on develop to this branch.

etoledom commented:
@SergioEstevao, ready with the merge from develop 👍

frosty commented:
Search bar styles look fixed now, and the `init` change for the media no results view is great!

I'm seeing one final issue, which is when I run `pod install`. At the end of the install process, I get:

```
[!] `<PBXSourcesBuildPhase UUID=`1D60588E0D05DD3D006BFB54`>` attempted to initialize an object with an unknown UUID. `5D8CBC471A6F47880081F4AE` for attribute: `files`. This can be the result of a merge and  the unknown UUID is being discarded.
```

Are either of you seeing this? I think it's usually due to a merge problem in the pbxproj file.

etoledom commented:
I'm checking it out, and for what I see, the problem was exactly what @frosty said.
` EditImageDetailsViewController` was removed in a previous commit but I merged it in the pbxproj file.

It was the merge conflict GitHub detected before but Xcode didn't detect it as a conflict.

My bad! 🙏
